File: Asgard_Test/tests_Heimdall/benchmarks/corpus/taint_js/tp_sa4_branch_join_unions.js
const { exec } = require('child_process');

// SA4 branch-join generalization: one arm assigns the tainted value, the
// other assigns a constant -- the join must UNION (taint on ANY reaching
// path survives), never intersect. This must still flag.
app.post('/ping', (req, res) => {
    const host = req.body.host;
    let cmd;
    if (req.query.flag) {
        cmd = host;
    } else {
        cmd = 'localhost';
    }
    exec('ping -c 1 ' + cmd, (err, stdout) => {
        res.send(stdout);
    });
});

File: Asgard_Test/tests_Heimdall/benchmarks/corpus/taint/tp_sa4_branch_join_unions.py
"""TP: SA4 branch-join generalization -- one arm assigns the tainted value,
the other assigns a constant. The join must UNION (taint on ANY reaching
path survives), never intersect. Must still flag."""

import subprocess


def run():
    host = request.args.get("host")
    flag = request.args.get("flag")
    if flag:
        cmd = host
    else:
        cmd = "localhost"
    subprocess.run("ping -c 1 " + cmd, shell=True)
